pkg/providers/cloudstack/validator.go

```diff
@@ -0,0 +1,197 @@
+	}
+
+	if err = v.validateMachineConfig(ctx, clusterSpec.datacenterConfig.Spec, controlPlaneMachineConfig); err != nil {
+		logger.V(1).Info("Control plane machine config validation failed.")
+		return err
+	}
+	if controlPlaneMachineConfig.Spec.Template != workerNodeGroupMachineConfig.Spec.Template {
+		if workerNodeGroupMachineConfig.Spec.Template.Value == "" {
+			return fmt.Errorf("worker CloudStackMachineConfig template is not set. Default template is not supported in CloudStack, please provide a template name")
+		}
+		if err = v.validateMachineConfig(ctx, clusterSpec.datacenterConfig.Spec, workerNodeGroupMachineConfig); err != nil {
+			logger.V(1).Info("Workload machine config validation failed.")
+			return err
+		}
+		if controlPlaneMachineConfig.Spec.Template != workerNodeGroupMachineConfig.Spec.Template {
+			return errors.New("control plane and worker nodes must have the same template specified")
+		}
+	}
+	logger.MarkPass("Control plane and Workload templates validated")
+
+	if etcdMachineConfig != nil {
+		if etcdMachineConfig.Spec.Template.Value == "" {
+			return fmt.Errorf("etcd CloudStackMachineConfig template is not set. Default template is not supported in CloudStack, please provide a template name")
+		}
+		if err = v.validateMachineConfig(ctx, clusterSpec.datacenterConfig.Spec, etcdMachineConfig); err != nil {
+			logger.V(1).Info("Etcd machine config validation failed.")
+			return err
+		}
+		if etcdMachineConfig.Spec.Template != controlPlaneMachineConfig.Spec.Template {
+			return errors.New("control plane and etcd machines must have the same template specified")
+		}
+	}
+
+	return nil
+}
+
+func (v *Validator) validateMachineConfig(ctx context.Context, datacenterConfigSpec anywherev1.CloudStackDatacenterConfigSpec, machineConfig *anywherev1.CloudStackMachineConfig) error {
+	domain := datacenterConfigSpec.Domain
+	zone := datacenterConfigSpec.Zone
+	account := datacenterConfigSpec.Account
+	if err := v.cmk.ValidateTemplatePresent(ctx, domain, zone, account, machineConfig.Spec.Template); err != nil {
+		return fmt.Errorf("error validating template: %v", err)
+	}
+	if err := v.cmk.ValidateServiceOfferingPresent(ctx, domain, zone, account, machineConfig.Spec.ComputeOffering); err != nil {
+		return fmt.Errorf("error validating template: %v", err)
+	}
+	if len(machineConfig.Spec.AffinityGroupIds) > 0 {
+		if err := v.cmk.ValidateAffinityGroupsPresent(ctx, domain, zone, account, machineConfig.Spec.AffinityGroupIds); err != nil {
+			return fmt.Errorf("error validating affinity group ids: %v", err)
+		}
+	}
+
+	return nil
+}
+
+func (v *Validator) validateControlPlaneHost(pHost *string) error {
+	_, port, err := net.SplitHostPort(*pHost)
+	if err != nil {
+		if strings.Contains(err.Error(), "missing port") {
+			port = controlEndpointDefaultPort
+			*pHost = fmt.Sprintf("%s:%s", *pHost, port)
+		} else {
+			return fmt.Errorf("cluster controlPlaneConfiguration.Endpoint.Host is invalid: %s (%s)", *pHost, err.Error())
```

Comment:
This function isn't operating on a `controlPlaneConfiguration.Endpoint.Host` so it shouldn't really add that context. The struct path context can be added in the caller instead.